File: 19. GraphBFSDFS.py
"""
Task: Implementation and Time Analysis of BFS and DFS Traversal of a Graph
"""
def bfs(adj, start):
    n = len(adj)
    visited = [False]*n
    queue = [start]
    visited[start] = True
    while queue:
        u = queue.pop(0)
        print(u, end=' ')
        for v in adj[u]:
            if not visited[v]:
                visited[v] = True
                queue.append(v)
def dfs_util(adj, u, visited):
    visited[u] = True
    print(u, end=' ')
    for v in adj[u]:
        if not visited[v]:
            dfs_util(adj, v, visited)
def dfs(adj, start):
    n = len(adj)
    visited = [False]*n
    dfs_util(adj, start, visited)
if __name__ == "__main__":
    import time
    n = 5
    adj = [
        [1,2], [0,3,4], [0,4], [1], [1,2]
    ]
    start = time.time()
    print("BFS:", end=' ')
    bfs(adj, 0)
    end = time.time()
    print(f"\nTime: {end-start:.6f}s")
    start = time.time()
    print("DFS:", end=' ')
    dfs(adj, 0)
    end = time.time()
    print(f"\nTime: {end-start:.6f}s")
"""
Time Complexity: O(V+E)
Space Complexity: O(V)
"""

File: 19. GraphBFSDFS.cpp
/*
Task: Implementation and Time Analysis of BFS and DFS Traversal of a Graph
*/
#include <iostream>
#include <vector>
#include <queue>
#include <ctime>
using namespace std;

void BFS(const vector<vector<int>>& adj, int start)
{
    int n = adj.size();
    vector<bool> visited(n, false);
    queue<int> q;
    q.push(start);
    visited[start] = true;
    while (!q.empty())
    {
        int u = q.front();
        q.pop();
        cout << u << " ";
        for (int v : adj[u])
        {
            if (!visited[v])
            {
                visited[v] = true;
                q.push(v);
            }
        }
    }
}
void DFSUtil(const vector<vector<int>>& adj, int u, vector<bool>& visited)
{
    visited[u] = true;
    cout << u << " ";
    for (int v : adj[u])
    {
        if (!visited[v])
            DFSUtil(adj, v, visited);
    }
}
void DFS(const vector<vector<int>>& adj, int start)
{
    int n = adj.size();
    vector<bool> visited(n, false);
    DFSUtil(adj, start, visited);
}
int main()
{
    int n = 5;
    vector<vector<int>> adj = {{1, 2}, {0, 3, 4}, {0, 4}, {1}, {1, 2}};
    clock_t start = clock();
    cout << "BFS: ";
    BFS(adj, 0);
    clock_t end = clock();
    cout << "\nTime: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    start = clock();
    cout << "DFS: ";
    DFS(adj, 0);
    end = clock();
    cout << "\nTime: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(V+E)
Space Complexity: O(V)
*/
